### PR TITLE
Revise Engineering Ladder

### PR DESCRIPTION
**Pull request by akubetin**
This is a work in progress, and so far I have only tackled most of the "Maker" roles.

In general, my focus has been on:
- consolidating redundant bullet points
- reorganizing expectations under separate categories (skills?)
- adding "Example behaviors" to each category to make the expectations feel more concrete (hopefully)
- differentiating more clearly between the roles

Some changes to note:
- Renamed the "Engineer" role to "Engineer I" (akin to Junior Engineer)
- Added a new role called "Engineer II" (akin to... Software Engineer, I guess? Whatever is after Junior Engineer)
- Added more information to the README about Makers vs Managers, as well as a high-level explanation of the ladder's purpose

Questions for reviewers:
- Does breaking expectations out by category make sense? Should the categories be different?
- Can you think of additional "example behaviors" for each role?
- Did I remove any bullet points from the original descriptions that you think are key?
- Did I add any bullet points that you hate?
- Can you think of any more to add?
- Should we add information about expected years of experience?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,20 +1,47 @@
-This repo contains a list of Engineering titles, roles, and associated responsibilities:
+The Mariana Tek Engineering Ladder defines the responsibilities and expectations of people working at various levels within the engineering team. It exists as a guide for engineers pursuing excellence and advancement, as well as a framework for managers to evaluate performance and identify candidates for promotion.
```
Comment (bashhack) [inline README.md]: Well phrased, and accurate to our discussions and intentions IMO!

File: README.md [hunk 2]
```diff
@@ -1,20 +1,47 @@
...
 
+Like many engineering teams, we outline two paths for career advancement: the management track and the technical track. Engineers pursuing the management track ("Managers") are interested in managing people and teams. Engineers pursuing the technical track ("Makers") are interested in honing their technical expertise and shaping the technical direction of the company.
+
+Framed another way by engineering team at [Chartbeat](https://engineering.chartbeat.com/2015/06/05/engineering-ladders/), the difference between Makers and Managers is how they answer this question:
+
+> “Do I want to build bigger and better systems, or do I want to manage bigger and better teams?”
+
+Every engineer starts their career as a maker. As you work your way from Engineer I to Engineer II to Senior Engineer, you are primarily focused on expanding your technical knowledge and mastering the skills required of a professional software engineer. When you're ready to move beyond Senior Engineer, you have the opportunity to specialize.
+
+Neither track is superior to the other, and engineers will often switch between them as they move through their careers.
```
Comment (bashhack) [inline README.md:26]: 👍 Great! Really glad we call this fluidity out, as I think the absence of such a remark may inadvertently give an impression of limitation or restriction.

File: README.md [hunk 3]
```diff
@@ -1,20 +1,47 @@
...
 
+Like many engineering teams, we outline two paths for career advancement: the management track and the technical track. Engineers pursuing the management track ("Managers") are interested in managing people and teams. Engineers pursuing the technical track ("Makers") are interested in honing their technical expertise and shaping the technical direction of the company.
+
+Framed another way by engineering team at [Chartbeat](https://engineering.chartbeat.com/2015/06/05/engineering-ladders/), the difference between Makers and Managers is how they answer this question:
+
+> “Do I want to build bigger and better systems, or do I want to manage bigger and better teams?”
+
+Every engineer starts their career as a maker. As you work your way from Engineer I to Engineer II to Senior Engineer, you are primarily focused on expanding your technical knowledge and mastering the skills required of a professional software engineer. When you're ready to move beyond Senior Engineer, you have the opportunity to specialize.
```
Comment (bashhack) [inline README.md]: I'm for the nomenclature here, but as per our discussions - there's just a gap between what we titling between systems. I don't think this is a hard block, because - frankly - our titles (as with many teams) have never cleanly aligned with any sort of rigid level stratification.

In as much as the `I` and `II` provide some space for us acknowledge junior, mid-level, and senior titles formally.  What discrepancy may exist on paper or Paylocity can be (ideally and I think in large part) addressed via other avenues of review.

File: README.md [hunk 4]
```diff
@@ -1,20 +1,47 @@
...
+
+Neither track is superior to the other, and engineers will often switch between them as they move through their careers.
 ### Makers
 
-* [CTO](makers/cto.md)
-* [Principal Engineer](makers/principal_engineer.md)
-* [Staff Engineer](makers/staff_engineer.md)
+* [Engineer I](makers/engineer_i.md)
```
Comment (bashhack) [inline README.md:30]: I like the flip here @akubetin, subtle but it's working for me 👍

File: makers/engineer_i.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+## Qualifications
+
+- Solid understanding of core CS concepts.
+- Familiarity with programming fundamentals: data structures, classes, debuggers, etc.
+- Connection to Mariana Tek's mission and values.
+
+## Responsibilities & Expectations
+
+<!-- TODO: Add summary -->
```
Comment (bashhack) [inline makers/engineer_i.md]: 👋 What blockers do we foresee to outlining these responsibilities?

File: makers/engineer_i.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+### Technical Skills
+
+- Complete small, well-defined engineering tasks within a reasonable time frame, working individually or pairing with others.
+- Write code, tests, and documentation consistent with team standards.
+- Collect and incorporate peer feedback (i.e. code reviews) into your work. Apply lessons learned to future projects.
+- Write tickets that contain enough detail to be picked up by other engineers.
+- Seek the right balance between persevering on a problem and asking for help.
+- Strive to improve your understanding of engineering principles, practices, and tools with the goal of becoming more self-sufficient.
+
+Example behaviors:
```
Comment (bashhack) [inline makers/engineer_i.md]: This recurring device of stating behaviors is great!

File: makers/engineer_ii.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
+Engineer II
+===========
+
+<!-- TODO: Add summary -->
+
+## Qualifications
+
+- Solid track record of writing clean, functional code with periodic guidance from other engineers.
```
Comment (bashhack) [inline makers/engineer_ii.md:7]: You've done I think an excellent job here articulating a set of qualifications that builds on top of Engineer I in a progressive fashion 💯

File: makers/senior_engineer.md [hunk 8]
```diff
@@ -1,35 +1,78 @@
...
+
+- Successfully performed the role of Engineer at Mariana Tek or equivalent role elsewhere.
+- Succeeded as an engineer when assigned projects of increasing complexity over time across multiple product domains.
+- Consistent track record of writing high-quality code with very little oversight beyond high-level direction.
+- Demonstrated ability to learn from mistakes.
+- Demonstrated ability to unblock themselves and others.
+- Communicates effectively with colleagues at all levels.
+## Responsibilities & Expectations
+
+<!-- TODO: Add summary -->
```
Comment (bashhack) [inline makers/senior_engineer.md]: I think there's something interesting happening here - as I am progressing through your proposed changes, I'm starting to question what would value a summary section at the top of the `Responsibilities & Expectations` section would have given the clarity and thoroughness of the content below within the sub-sections `Technical Skills`, `Business Knowledge`, `Communication`, et cetera?

My vote here is that we forego a summary section in the interest of:

- Getting to merge with fewer blocks
- Keeping the content actionable and clear
- Leaning on the clarity of the surrounding text

TLDR - I simply don't think there's a huge amount of value in the summary compared to the vast value presented already in the sub-sections.

File: makers/staff_engineer.md [hunk 9]
```diff
@@ -1,47 +1,78 @@
...
-* Challenge their team to demonstrate greater sense of urgency
-* Model bringing solutions to problems (not surfacing problems alone) and teach others on their team how to do the same
-* Be active in the technology community, positively contribute to Mariana Tek’s reputation, and find great candidates to grow the Mariana Tek team
-* Develop and constantly refine technical and team fit assessments (coding exercises, interview questions, etc) for interviews
-* Ensure code/design/architecture reviews happen on teams and that work follows established best practices and standards
-* Coordinate evaluation of tools and frameworks for the engineering team
-* Take responsibility for actions of their team, good and bad
+<!-- TODO: Add summary -->
+
+## Qualifications
```
Comment (bashhack) [inline makers/staff_engineer.md:4]: I feel this captures the role well - looking to @cathrynsupko here as well for a gut check. I think the balance between what we are tasked with both technically and non-technically feels pretty accurate to me, TBH.

File: makers/staff_engineer.md [hunk 10]
```diff
@@ -1,47 +1,78 @@
...
+- Evaluate new technologies and tools for the engineering team to adopt.
+
+Example behaviors:
+- Worked with the PM to break a large feature initiative into smaller tickets before grooming
+- Planned and executed a major version upgrade of an important package
+- Successfully led the team-wide adoption of a new linting tool
+- Flagged a feature that posed a performance risk for a large new client and outlined a plan to refactor it
+
+---
+### Business Knowledge
```
Comment (bashhack) [inline makers/staff_engineer.md:35]: I think I want more here - can I suggest that (in my experience) that stating our responsibility to thoroughly understand the risk vs value tradeoff of a product effort is essential at this level.

Maybe something like 'works with Product teams to understand intersection of engineering effort and product initiative'?

File: makers/staff_engineer.md [hunk 11]
```diff
@@ -1,47 +1,78 @@
...
+- Steps up to take leadership of a project or team when needed.
+- Respected by engineers across the company as a resource for technical knowledge, leadership, and mentoring.
+- Trusted by company leaders to contribute strong technical guidance to collaborative decisions.
+- Demonstrated ability to take a nebulous project, perform additional research, architect a solution, break the work into subtasks, and coordinate delivery quickly and efficiently among multiple engineers.
+- Demonstrated ability to diagnose, fix, and coordinate deployment of bug fixes -- and to do this extremely quickly in emergency situations.
+## Responsibilities & Expectations
+
+<!-- TODO: Add summary -->
+
+### Technical Skills
```
Comment (bashhack) [inline makers/staff_engineer.md:17]: You know what @akubetin - I just realized that from Engineer I on up, you've never focused too heavily on the lingo of our stack and I think that's a great move! I think this speaks to our intent to provide engineers across the stack opportunity for growth and clear benchmarks.

File: README.md (hunk 3 above)
Comment (akubetin) [reply README.md]: Would you prefer to change `Engineer I` to `Junior Engineer` and `Engineer II` to `Mid-Level Engineer` or `Engineer`?

File: README.md (hunk 4 above)
Comment (akubetin) [reply README.md:30]: I think it makes it more readable, but it also inverts the ladder. I'd be happy to revert the order if people feel strongly about it.

File: makers/engineer_i.md (hunk 5 above)
Comment (akubetin) [reply makers/engineer_i.md]: Mainly how slowly I write.

File: makers/senior_engineer.md (hunk 8 above)
Comment (akubetin) [reply makers/senior_engineer.md]: My general idea for the summary section would be a sentence or two outlining the high-level function/goal of each role, more articulately than what I've noted below:

- Engineer I: A new engineer focused on learning the ropes of software development, learning to write code consistent with team standards, growing their development skills
- Engineer II: A solid contributor, often self-sufficient & becoming more so, developing expertise in a particular domain, contributing more to the technical design process with guidance from more senior engineers
- Senior Engineer: A strong independent contributor, developing your tech leadership skills, handling end-to-end responsibility for projects of increasing complexity
- Staff Engineer: A team leader? A cross-team leader? Provide high-level technical guidance across your team / teams, help colleagues build and produce better outcomes
- Principal Engineer: Primarily focusing on long-term architectural and technology needs for parts of the business
- CTO: An organizational leader, responsible for setting company-wide technical direction

File: makers/staff_engineer.md (hunk 10 above)
Comment (akubetin) [reply makers/staff_engineer.md:35]: Good suggestion, will do!

File: README.md [hunk 12]
```diff
@@ -1,20 +1,49 @@
...
 
+Like many engineering teams, we outline two paths for career advancement: the management track and the technical track. Engineers pursuing the management track ("Managers") are interested in managing people and teams. Engineers pursuing the technical track ("Makers") are interested in honing their technical expertise and shaping the technical direction of the company.
+
+Framed another way by the engineering team at [Chartbeat](https://engineering.chartbeat.com/2015/06/05/engineering-ladders/), the difference between Makers and Managers is how they answer this question:
+
+> “Do I want to build bigger and better systems, or do I want to manage bigger and better teams?”
+
+Every engineer starts their career as a maker. As you work your way from Engineer I to Engineer II to Senior Engineer, you are primarily focused on expanding your technical knowledge and mastering the skills required of a professional software engineer. When you're ready to move beyond Senior Engineer, you have the opportunity to specialize.
```
Comment (tracyalison11) [inline README.md]: Confused by this portion because I am not senior but am specialized in front-end

Comment (akubetin) [reply README.md]: I will rephrase to clarify that I mean choosing between the technical track + managerial track. Good note!

File: README.md (hunk 1 above)
Comment (krisfields) [reply README.md]: It's also useful as a guide for hiring.  IMO the appropriate section should be read by each interviewer before giving an interview so as to provide a framework around what questions are appropriate and what skills/qualifications we're looking for for that role.  Doing so would also help eliminate natural biases.

File: README.md (hunk 12 above)
Comment (krisfields) [reply README.md]: 👍  This section is really well written and so much better than before.  @tracyalison11 makes a great call out about specialization though.  I think all of our engineers are somewhat specialized (mobile, frontend, backend, develops, ect...) and we seem to be trending towards more and more specialized teams.  If anything it almost seems like our expectations would be the reverse where as you become more and more senior you'd be expected to branch out and learn more about our entire stack vs the part you're initially comfortable in.  But tbh I'm not really sure what the expectation is there.

File: README.md (hunk 3 above)
Comment (tracyalison11) [reply README.md]: I'm not sure I fully understand Marc's comments and hope he will be able to elaborate. Did you mean that there has always been a level of ambiguity with titles for better or for worse? Are you suggesting not having these titles so much an reinforcing them through these "other avenues of review"?

FWIW, I prefer `Engineer I` and `Engineer II`. I am curious what @Asilo5 prefers?

File: makers/engineer_ii.md (hunk 7 above)
Comment (tracyalison11) [reply makers/engineer_ii.md:7]: Agree - I think these all make sense and provide great examples. I hope others on the team will find inspiration from examples like:

> Demoed a new feature at Product Review and explained how it helps meet business needs

I hadn't thought how I could better contextualize these demos and will be now

File: README.md (hunk 3 above)
Comment (Asilo5) [reply README.md]: I like `Engineer I` and `Engineer II`. I think it states what stage the developer is or will be starting at pretty well, and the growth from that can be understood throughout the time.

File: makers/staff_engineer.md (hunk 10 above)
Comment (thomasmatecki) [reply makers/staff_engineer.md:35]: Is there any value in using language that highlights a difference between _Industry_ and _Business_ knowledge, or simply a call out that one or both is more important at certain levels or on certain career tracks.

As an example, perhaps there's a large focus on _Industry knowledge_ at the SE 1 & 2 levels, but expectations increase in the _Business knowledge_ as folks approach the Staff/ME/Principle level/Director?

File: README.md (hunk 3 above)
Comment (bashhack) [reply README.md]: Definitely a strict preference for `I` and `II` here - would push back against using junior or mid-level at all costs personally.

@tracyalison11 - my points around title discrepancy here are twofold, really:

- (1) just acknowledging that titles were all over the place in the early days of Mariana Tek (not uncommon of course in our industry) 
- (2) emphasizing that while a level-system (whether its the one we're all in agreement on here with `I` and `II`, etc or something like `junior-`, `mid-level-`, etc in other companies) will hopefully provide additional clarity for both existing and future team members, we may still expect to see a gap in how titling is presented across systems (i.e., one's title may appear in a corporate HR system differently than how it might appear in an employee review document, etc)

The takeaway here ideally is that even if the Paylocity system of record or Workplace may not have the level of granularity we'd hope for - these titles hopefully become the standard by which we use in reviews, promotions, etc. within the Mariana Tek Engineering team(s).

Hopefully that makes sense to all!

File: makers/senior_engineer.md (hunk 8 above)
Comment (bashhack) [reply makers/senior_engineer.md]: 🤔 Okay - this is growing on me! I trust your gut here - can go either way of course, I do like your proposals and think we can get good value of these summarizations 👍

File: makers/staff_engineer.md (hunk 10 above)
Comment (bashhack) [reply makers/staff_engineer.md:35]: Ooh - I like this idea @thomasmatecki ! That phrasing is a nice distinction 💯

Comment (akubetin) [reply makers/staff_engineer.md:35]: That's a really good note, @thomasmatecki! I will make some changes.

File: README.md (hunk 2 above)
Comment (tguidon) [reply README.md:26]: 👍 plus one on calling out the fluidity in paths

File: README.md (hunk 3 above)
Comment (tguidon) [reply README.md]: @bashhack do we know if there are plans to align the titles across TSG? Even if just across the boutique vertical? While it may not be possible to align titles in every system (Paylocity for example), I see value in aligning our titles across the Boutique brands.

File: README.md [hunk 13]
```diff
@@ -1,20 +1,103 @@
...
 
-### Roles
+## Roles
 
 * [Technical Lead](roles/technical_lead.md)
 * [Directly Responsible Individual (DRI)](roles/directly_responsible_individual.md)
+
+## For New Employees
+
+If you are new to the team, please don't panic while reading the expectations for engineers at your level! Our codebase is huge and complicated, and it's normal to spend 6-12 months on the team before you feel like you know the lay of the land. You will contribute code before then, of course! We do a lot of pair programming and skill-sharing during the onboarding process, and you'll start off with small, well-defined tickets to build up your confidence. Ramping up to your full capabilities will take a while -- and that's okay!
```
Comment (tguidon) [inline README.md:50]: 👍 This is a great piece to read for new employees, it defitintely can be overwhelming looking at our codebase on day one.

File: makers/engineer_i.md [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+- Collect and incorporate peer feedback (i.e. code reviews) into your work. Apply lessons learned to future projects.
+- Write tickets that contain enough detail to be picked up by other engineers.
+- Seek the right balance between persevering on a problem and asking for help.
+- Strive to improve your understanding of engineering principles, practices, and tools with the goal of becoming more self-sufficient.
+
+[Example Behaviors](/README.md#example-behaviors):
+- Identified, documented, and fixed a small bug
+- Exposed a new field on an existing API endpoint
+- Paired with another developer to refactor an Ember route
+- Debugged and fixed failing tests in your PR before seeking code review
```
Comment (tguidon) [inline makers/engineer_i.md]: Noting we have an API and Web FE example. If we wanted to add an example for mobile roles, a good example could be:

- Created a new view that was added to a scrolling view in our mobile apps.

File: makers/senior_engineer.md [hunk 15]
```diff
@@ -1,35 +1,76 @@
...
+### [Leadership & Teamwork](#leadership--teamwork)
+
+- Set and maintain high expectations for yourself and others.
+- Nurture an empathetic, no-blame culture on your team.
+- Prioritize work that unblocks your teammates.
+- Provide constructive, actionable feedback to your colleagues in an empathetic way.
+- Suggest solutions to problems you notice with our product and processes.
+- Notice if morale is low, and work with the team to identify and address the causes.
+- Coach junior engineers via pairing and code reviews.
+- Volunteer to serve as a formal mentor for junior engineers.
```
Comment (tguidon) [inline makers/senior_engineer.md]: Do we want to use "junior" here or should we use verbiage that reflects the updates we made to the ladder? (Engineer I and II). Not sure what the best phrasing would be in the latter.

"Coach engineers starting their career via pairing and code reviews."?

**Review by bashhack — APPROVED** [review]
Thank you for your effort in updating this document and making it reflect our team and our values!

File: README.md (hunk 13 above)
Comment (Asilo5) [reply README.md:50]: Being currently in this spot, it is helpful and encouraging to see it in writing. Great that you added that in!
